### PR TITLE
Destroy ipsets prefixed `weave-` only (fixes issue #2751)

### PR DESCRIPTION
**Pull request by stephan2012**
Changes to destroy weave's ipsets only so that it can co-exist with other tools that use ipsets (e.g., keepalived-vip).

### CODE REVIEW

File: npc/ipset/ipset.go [hunk 1]
```diff
@@ -35,48 +39,77 @@ func New() Interface {
...
+	err, _ := doExec("destroy")
+	return err
+}
+
+// Fetch a list of all existing sets
+func (i *ipset) List(prefix string) ([]Name, error) {
+	err, output := doExec("list","-name","-output","plain")
+
+	var selected []Name
+	if err == nil && len(output) > 0 {
```
Comment (bboreham) [inline npc/ipset/ipset.go]: The Go idiom is to test for error and return, thus leaving the happy path un-indented.

File: npc/ipset/ipset.go [hunk 2]
```diff
@@ -35,48 +39,77 @@ func New() Interface {
...
+		output = bytes.TrimRight(output, "\n")
+		sets := strings.Split(string(output[:]), "\n")
```
Comment (bboreham) [inline npc/ipset/ipset.go]: why `[:]` ?

File: npc/ipset/ipset.go [hunk 3]
```diff
@@ -35,48 +39,77 @@ func New() Interface {
...
+
+		plen := len(prefix)
+		for _, v := range sets {
+			if (plen <= len(v)) && (prefix == v[:len(prefix)]) {
```
Comment (bboreham) [inline npc/ipset/ipset.go]: `strings.HasPrefix(v, prefix)`

File: npc/ipset/ipset.go [hunk 4]
```diff
@@ -35,48 +39,77 @@ func New() Interface {
...
+
+	}
+
+	return selected, err
 }
 
-func doExec(args ...string) error {
-	if output, err := exec.Command("ipset", args...).CombinedOutput(); err != nil {
-		return errors.Wrapf(err, "ipset %v failed: %s", args, output)
+func doExec(args ...string) (error, []byte) {
```
Comment (bboreham) [inline npc/ipset/ipset.go]: Also it gave a `gofmt` error:

npc/ipset/ipset.go:107:1: error should be the last type when returning multiple items

**Comment by bboreham** [comment]
Thanks for the contribution, @stephan2012 ! 

I've made some comments in a review.

File: npc/ipset/ipset.go [hunk 5]
```diff
@@ -35,48 +39,77 @@ func New() Interface {
...
+	return err
+}
+
+// Fetch a list of all existing sets
+func (i *ipset) List(prefix string) ([]Name, error) {
+	err, output := doExec("list","-name","-output","plain")
+
+	var selected []Name
+	if err == nil && len(output) > 0 {
+		output = bytes.TrimRight(output, "\n")
```
Comment (stephan2012) [inline npc/ipset/ipset.go]: Without stripping the final newline we have an empty element in the list of prefixes.

Comment (bboreham) [inline npc/ipset/ipset.go]: which will then fail the `HasPrefix` test

File: prog/weave-npc/main.go [hunk 6]
```diff
@@ -80,14 +80,22 @@ func resetIPTables(ipt *iptables.IPTables) error {
...
 		return err
 	}
 
-	if err := ips.DestroyAll(); err != nil {
-		return err
+	 common.Log.Debugf("Got list of ipsets: %v", sets)
+
+	for _, s := range sets {
+		common.Log.Debugf("Destroying ipsets '%s'", string(s))
+		if err := ips.Destroy(s); err != nil {
```
Comment (stephan2012) [inline prog/weave-npc/main.go]: I've tested it and ipset does not require flushing first:

```
# ipset --version
ipset v6.19, protocol version: 6

# ipset create foo bitmap:ip range 192.168.1.0/24  
# ipset add foo 192.168.1.11
# ipset add foo 192.168.1.12
# ipset add foo 192.168.1.13
# ipset destroy foo
```

File: npc/ipset/ipset.go (hunk 3 above)
Comment (stephan2012) [inline npc/ipset/ipset.go]: Ok, clearly looks better.

File: npc/ipset/ipset.go (hunk 4 above)
Comment (stephan2012) [inline npc/ipset/ipset.go]: That way you have execs in different places.

I've haven't seen any error when compiling it. The patched npc container is actually working in my Kubernetes setup. Maybe a Go version thing?

**Comment by stephan2012** [comment]
@bboreham: Thanks for your feedback. The idiomatic stuff is related to me being a newbie in Go…

What are the expectations regarding this PR?

File: npc/ipset/ipset.go (hunk 4 above)
Comment (bboreham) [inline npc/ipset/ipset.go]: the error comes when you run `gofmt -s -w npc/ipset/ipset.go`, which we do as part of the CI

**Comment by bboreham** [comment]
If you fix up the points we can merge it.  Or someone else will fix them up.

Thanks!

One more thought: it would be good to have a test for the extended functionality, e.g. create an `ipset` before running, then check it's still there afterwards, in `test/840_weave_kube_3_test.sh`.  But to execute that test you'd need to set up 3 VMs with Kubernetes installed, which is quite a lot of faff.  (We have scripts to do it, but they have dependencies, etc.)

**Comment by bboreham** [comment]
I made the changes that I suggested, and added a test: https://github.com/weaveworks/weave/compare/stephan2012-master

CI build passed at https://circleci.com/gh/weaveworks/weave/8207
I also checked that the test fails on existing 1.9: https://circleci.com/gh/weaveworks/weave/8208

**Comment by brb** [comment]
@bboreham Could you open PR with your changes?

**Comment by marccarre** [comment]
Closing this PR as #2846 has been merged instead, but thanks again for the contribution @stephan2012!

**Pull request closed without merging**